rbd: free snapshot resources after allocation

nixpanic opened the pull request:
Just like GenVolFromVolID() the genSnapFromSnapID() function can return
a snapshot. There is no need to allocated an empty snapshot and pass
that to the genSnapFromSnapID() function.

With this change, it is visible that not all snapshot objects are free'd correctly after they were allocated.
It is possible that some connections to the Ceph cluster were never
closed. This does not need to be a noticeable problem, as connections
are re-used where possible, but it isn't clean either.

nixpanic committed:
cleanup: do not pass an empty snapshot to genSnapFromSnapID()

Just like GenVolFromVolID() the genSnapFromSnapID() function can return
a snapshot. There is no need to allocated an empty snapshot and pass
that to the genSnapFromSnapID() function.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
rbd: free snapshot resources after allocation

Not all snapshot objects are free'd correctly after they were allocated.
It is possible that some connections to the Ceph cluster were never
closed. This does not need to be a noticeable problem, as connections
are re-used where possible, but it isn't clean either.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

Madhu-1 reviewed (dismissed):
LGTM

nixpanic commented:
@mergifyio queue

nixpanic commented:
oh no, interesting failure!
```
  I0321 15:32:43.841709       1 utils.go:198] ID: 40 Req-ID: 0001-0024-4a93465d-9d6b-49d5-a38e-8fb38ed70ee8-0000000000000002-1593e684-c8bd-4080-8910-009e5519ec7b GRPC call: /csi.v1.Controller/DeleteSnapshot
  I0321 15:32:43.841827       1 utils.go:199] ID: 40 Req-ID: 0001-0024-4a93465d-9d6b-49d5-a38e-8fb38ed70ee8-0000000000000002-1593e684-c8bd-4080-8910-009e5519ec7b GRPC request: {"secrets":"***stripped***","snapshot_id":"0001-0024-4a93465d-9d6b-49d5-a38e-8fb38ed70ee8-0000000000000002-1593e684-c8bd-4080-8910-009e5519ec7b"}
  E0321 15:32:43.841920       1 rbd_util.go:963] ID: 40 Req-ID: 0001-0024-4a93465d-9d6b-49d5-a38e-8fb38ed70ee8-0000000000000002-1593e684-c8bd-4080-8910-009e5519ec7b error decoding snapshot ID (failed to decode CSI identifier, string underflow) ()
  E0321 15:32:43.841956       1 utils.go:203] ID: 40 Req-ID: 0001-0024-4a93465d-9d6b-49d5-a38e-8fb38ed70ee8-0000000000000002-1593e684-c8bd-4080-8910-009e5519ec7b GRPC error: rpc error: code = Internal desc = failed to decode CSI identifier, string underflow
```

iPraveenParihar commented on internal/rbd/rbd_util.go:
```diff
@@ -951,52 +951,51 @@ func (ri *rbdImage) checkImageChainHasFeature(ctx context.Context, feature uint6
-
-	rbdSnap.VolID = snapshotID
+	rbdSnap := &rbdSnapshot{}
 
```
@nixpanic, forgot adding?
```go
rbdSnap.VolID = snapshotID
```

Madhu-1 replied:
how did i miss this one 🤦🏻

nixpanic replied:
Oh right, that is done a few lines  further below. Thanks for checking!

iPraveenParihar reviewed (changes requested):
^^ requested changes in above comment.

nixpanic commented:
@mergifyio queue

nixpanic commented:
ouch, still an issue
```
  I0322 09:47:28.500766       1 utils.go:198] ID: 37 Req-ID: 0001-0024-5536a82c-233b-4817-9b07-84c898f5a1ef-0000000000000002-1e3f3f48-a692-491a-a2c5-5c918bd8519f GRPC call: /csi.v1.Controller/DeleteSnapshot
  I0322 09:47:28.500918       1 utils.go:199] ID: 37 Req-ID: 0001-0024-5536a82c-233b-4817-9b07-84c898f5a1ef-0000000000000002-1e3f3f48-a692-491a-a2c5-5c918bd8519f GRPC request: {"secrets":"***stripped***","snapshot_id":"0001-0024-5536a82c-233b-4817-9b07-84c898f5a1ef-0000000000000002-1e3f3f48-a692-491a-a2c5-5c918bd8519f"}
  I0322 09:47:28.502745       1 omap.go:89] ID: 37 Req-ID: 0001-0024-5536a82c-233b-4817-9b07-84c898f5a1ef-0000000000000002-1e3f3f48-a692-491a-a2c5-5c918bd8519f got omap values: (pool="replicapool", namespace="", name="csi.snap.1e3f3f48-a692-491a-a2c5-5c918bd8519f"): map[csi.imageid:1ba6623dce46 csi.imagename:csi-snap-1e3f3f48-a692-491a-a2c5-5c918bd8519f csi.snapname:snapshot-e132b1c3-00f5-4cef-bd8b-10b7633f996c csi.source:csi-vol-f1e770d2-e32b-437b-ac0f-7a29dc13bca0 csi.volume.owner:rbd-8829]
  E0322 09:47:28.507028       1 utils.go:220] panic occurred: runtime error: invalid memory address or nil pointer dereference
  goroutine 142 [running]:
  runtime/debug.Stack()
  	/usr/local/go/src/runtime/debug/stack.go:24 +0x5e
  runtime/debug.PrintStack()
  	/usr/local/go/src/runtime/debug/stack.go:16 +0x13
  github.com/ceph/ceph-csi/internal/csi-common.panicHandler.func1()
  	/go/src/github.com/ceph/ceph-csi/internal/csi-common/utils.go:221 +0xca
  panic({0x1ea8140?, 0x3a71420?})
  	/usr/local/go/src/runtime/panic.go:914 +0x21f
  github.com/ceph/ceph-csi/internal/rbd.generateVolFromSnap(...)
  	/go/src/github.com/ceph/ceph-csi/internal/rbd/snapshot.go:103
  github.com/ceph/ceph-csi/internal/rbd.cleanUpImageAndSnapReservation({0x27bfcf0, 0xc0009631d0}, 0x0, 0x3a71d90?)
  	/go/src/github.com/ceph/ceph-csi/internal/rbd/controllerserver.go:1500 +0x6d
  github.com/ceph/ceph-csi/internal/rbd.(*ControllerServer).DeleteSnapshot(0xc000aa2280, {0x27bfcf0, 0xc0009631d0}, 0xc00093d300)
  	/go/src/github.com/ceph/ceph-csi/internal/rbd/controllerserver.go:1445 +0x7f5
  github.com/container-storage-interface/spec/lib/go/csi._Controller_DeleteSnapshot_Handler.func1({0x27bfcf0, 0xc0009631d0}, {0x2090080?, 0xc00093d300})
  	/go/src/github.com/ceph/ceph-csi/vendor/github.com/container-storage-interface/spec/lib/go/csi/csi.pb.go:6749 +0x72
  github.com/ceph/ceph-csi/internal/csi-common.panicHandler({0x27bfcf0?, 0xc0009631d0?}, {0x2090080?, 0xc00093d300?}, 0xc000960360?, 0xc000978020?)
  	/go/src/github.com/ceph/ceph-csi/internal/csi-common/utils.go:226 +0x77
  github.com/ceph/ceph-csi/internal/csi-common.NewMiddlewareServerOption.ChainUnaryServer.func2.1({0x27bfcf0?, 0xc0009631d0?}, {0x2090080?, 0xc00093d300?})
  	/go/src/github.com/ceph/ceph-csi/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:48 +0x45
  github.com/ceph/ceph-csi/internal/csi-common.logGRPC({0x27bfcf0, 0xc0009631d0}, {0x2090080?, 0xc00093d300?}, 0x1d76040?, 0xc00093d340)
  	/go/src/github.com/ceph/ceph-csi/internal/csi-common/utils.go:201 +0x183
  github.com/ceph/ceph-csi/internal/csi-common.NewMiddlewareServerOption.ChainUnaryServer.func2.1({0x27bfcf0?, 0xc0009631d0?}, {0x2090080?, 0xc00093d300?})
  	/go/src/github.com/ceph/ceph-csi/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:48 +0x45
  github.com/ceph/ceph-csi/internal/csi-common.contextIDInjector({0x27bfcf0, 0xc000962de0}, {0x2090080, 0xc00093d300}, 0xc000954c90?, 0xc00093d380)
  	/go/src/github.com/ceph/ceph-csi/internal/csi-common/utils.go:189 +0x132
  github.com/ceph/ceph-csi/internal/csi-common.NewMiddlewareServerOption.ChainUnaryServer.func2({0x27bfcf0, 0xc000962de0}, {0x2090080, 0xc00093d300}, 0xc000630cc0, 0x1ea8f20?)
  	/go/src/github.com/ceph/ceph-csi/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:53 +0x135
  github.com/container-storage-interface/spec/lib/go/csi._Controller_DeleteSnapshot_Handler({0x21482a0?, 0xc000aa2280}, {0x27bfcf0, 0xc000962de0}, 0xc000953480, 0xc000480ae0)
  	/go/src/github.com/ceph/ceph-csi/vendor/github.com/container-storage-interface/spec/lib/go/csi/csi.pb.go:6751 +0x135
  google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001eb800, {0x27bfcf0, 0xc000962d50}, {0x27c85e0, 0xc0005831e0}, 0xc000974000, 0xc000480cc0, 0x3a88a78, 0x0)
  	/go/src/github.com/ceph/ceph-csi/vendor/google.golang.org/grpc/server.go:1386 +0xe23
  google.golang.org/grpc.(*Server).handleStream(0xc0001eb800, {0x27c85e0, 0xc0005831e0}, 0xc000974000)
  	/go/src/github.com/ceph/ceph-csi/vendor/google.golang.org/grpc/server.go:1797 +0x100c
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
  	/go/src/github.com/ceph/ceph-csi/vendor/google.golang.org/grpc/server.go:1027 +0x8b
  created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 123
  	/go/src/github.com/ceph/ceph-csi/vendor/google.golang.org/grpc/server.go:1038 +0x135
  E0322 09:47:28.507246       1 utils.go:203] ID: 37 Req-ID: 0001-0024-5536a82c-233b-4817-9b07-84c898f5a1ef-0000000000000002-1e3f3f48-a692-491a-a2c5-5c918bd8519f GRPC error: rpc error: code = Internal desc = panic runtime error: invalid memory address or nil pointer dereference
```

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1416,8 +1422,8 @@ func (cs *ControllerServer) DeleteSnapshot(
 
-	rbdSnap := &rbdSnapshot{}
-	if err = genSnapFromSnapID(ctx, rbdSnap, snapshotID, cr, req.GetSecrets()); err != nil {
+	rbdSnap, err := genSnapFromSnapID(ctx, snapshotID, cr, req.GetSecrets())
```
in case of err , rbdSnap is still used in `cleanUpImageAndSnapReservation` which is causing issues.

nixpanic replied:
yes, I noticed. Super ugly :sob:

nixpanic replied:
Tested manually with returning a partial generated rbdSnapshot for now. It is not as clean as I would like, but it is still better than what was done before.

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.29

nixpanic commented:
@mergifyio rebase

initial rbd tests related to snapshots have passed [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/job/mini-e2e_k8s-1.29/142/display/redirect)

nixpanic commented:
oh, wow, after the rebase this suddenly pops up:
```
internal/rbd/controllerserver.go:310:1: cyclomatic complexity 21 of func `(*ControllerServer).CreateVolume` is high (> 20) (gocyclo)
```

nixpanic committed:
rbd: let parseVolCreateRequest() return a connected rbdVolume

By returning a connected rbdVolume in parseVolCreateRequest(), the
CreateVolume() function can be simplified a little. There is no need to
call the additional Connect() and detect failures with it.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1009,42 +1012,42 @@ func genSnapFromSnapID(
 	err = updateSnapshotDetails(rbdSnap)
 	if err != nil {
-		return fmt.Errorf("failed to update snapshot details for %q: %w", rbdSnap, err)
+		return nil, fmt.Errorf("failed to update snapshot details for %q: %w", rbdSnap, err)
```
need to return `rbdSnap` which is required  for the case of `ErrImageNotFound` error message

nixpanic replied:
Now returning `rbdSnap` on all but the initial errors. More manual testing suggests this should be ok.

nixpanic committed:
cleanup: reformat generateVolFromSnap() to rbdSnapshot.toVolume()

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
rbd: add extra logging while cleaning up snapshots

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
:sob: it's Tuesday again and pulling container-images in the CI is problematic...

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.29

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.29

nixpanic commented:
@mergifyio rebase
